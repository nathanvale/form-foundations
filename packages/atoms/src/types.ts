import {StandardProps} from '@material-ui/core'
import {
  FormControlClassKey,
  FormControlProps,
} from '@material-ui/core/FormControl'
import {FormHelperTextProps} from '@material-ui/core/FormHelperText'
import {FormLabelProps} from '@material-ui/core/FormLabel'
import {RadioProps as MUIRadioProps} from '@material-ui/core/Radio'
import {RadioGroupProps} from '@material-ui/core/RadioGroup'
import {TextFieldProps as MUITextFieldProps} from '@material-ui/core/TextField'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export type SelectFieldNativeProps = {options: any} & TextFieldProps

export type TextFieldProps = {
  helperText?: string | undefined
} & MUITextFieldProps

export type RadioProps = {
  error?: boolean
  'data-id'?: string
} & Omit<
  MUIRadioProps,
  'color' | 'classes' | 'disableRipple' | 'checkedIcon' | 'icon' | 'id'
>

export interface RadioGroupFieldProps
  extends StandardProps<
    FormControlProps,
    RadioGroupFieldClassKey,
    'onChange' | 'defaultValue'
  > {
  error?: boolean
  fullWidth?: boolean
  required?: boolean
  label?: React.ReactNode
  FormLabelProps?: Partial<FormLabelProps>
  id?: string
  onChange?: (event: React.ChangeEvent<{}>, value: string) => void
  value?: string
  name?: string
  placeholder?: string
  RadioGroupProps?: Partial<RadioGroupProps>
  helperText?: React.ReactNode
  FormHelperTextProps?: Partial<FormHelperTextProps>
}

export interface RadioGroupFieldInnerProps extends RadioGroupFieldProps {
  handleChange: any
  state: any
  helperTextId: string
}

export type RadioGroupFieldClassKey = FormControlClassKey
