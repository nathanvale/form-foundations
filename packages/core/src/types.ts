import {
  RadioGroupFieldProps as RadioGroupFieldAtomProps,
  SelectFieldNativeProps as SelectFieldNativeAtomProps,
  TextFieldProps as TextFieldAtomProps,
} from '@form-foundations/atoms';
import {
  FieldConfig as FormikFieldConfig,
  FieldProps,
  GenericFieldHTMLAttributes,
  FormikProps,
  FormikSharedConfig,
  FormikState,
  FormikActions,
  FormikHandlers,
  FormikComputedProps,
  FormikRegistration,
  FormikErrors,
} from 'formik';
import * as React from 'react';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type FieldPropsInjected<Props = {}, Values = {}> = FieldProps<Values> &
  Props;

/**
 * Note: These typings could be more restrictive, but then it would limit the
 * reusability of custom <Field/> components.
 *
 * @example
 * interface MyProps {
 *   ...
 * }
 *
 * export const MyInput: React.SFC<MyProps & FieldProps> = ({
 *   field,
 *   form,
 *   ...props
 * }) =>
 *   <div>
 *     <input {...field} {...props}/>
 *     {form.touched[field.name] && form.errors[field.name]}
 *   </div>
 */
export interface FieldProps<V = any> {
  field: {
    /** Classic React change handler, keyed by input name */
    onChange: (e: React.ChangeEvent<any>) => void;
    /** Mark input as touched */
    onBlur: (e: any) => void;
    /** Value of the input */
    value: any;
    /* name of the input */
    name: string;
  };
  form: FormikProps<V>; // if ppl want to restrict this for a given form, let them.
}

export interface FieldConfig {
  /**
   * Field component to render. Can either be a string like 'select' or a component.
   */
  component?:
    | string
    | React.ComponentType<FieldProps<any>>
    | React.ComponentType<void>;

  /**
   * Render prop (works like React router's <Route render={props =>} />)
   */
  render?: ((props: FieldProps<any>) => React.ReactNode);

  /**
   * Children render function <Field name>{props => ...}</Field>)
   */
  children?: ((props: FieldProps<any>) => React.ReactNode) | React.ReactNode;

  /**
   * Validate a single field value independently
   */
  validate?: ((value: any) => string | Promise<void> | undefined);

  /**
   * Field name
   */
  name: string;

  /** HTML input type */
  type?: string;

  /** Field value */
  value?: any;

  /** Inner ref */
  innerRef?: (instance: any) => void;
}

export type FieldAttributes<T> = GenericFieldHTMLAttributes & FieldConfig & T;

export type RadioGroupFieldProps =
  | Omit<
      Partial<RadioGroupFieldAtomProps>,
      'error' | 'onChange' | 'value' | 'onBlur'
    >
  | any;

export type TextFieldProps =
  | Omit<Partial<TextFieldAtomProps>, 'error' | 'onChange' | 'value' | 'onBlur'>
  | any;

export type TextFieldNormalizeProps = {
  normalize?: (value: string, previousValue: string) => string;
} & TextFieldProps;

export type SelectFieldNativeProps =
  | Omit<
      Partial<SelectFieldNativeAtomProps>,
      'error' | 'onChange' | 'value' | 'onBlur'
    >
  | any;

// Extended from Formik
export interface FFConfig<Values> extends FormikSharedConfig {
  component?: React.ComponentType<FormikProps<Values>> | React.ReactNode;
  render?: ((props: FFProps<Values>) => React.ReactNode);
  children?: ((props: FFProps<Values>) => React.ReactNode) | React.ReactNode;
  initialValues: Values;
  onReset?: (values: Values, formikActions: FFActions<Values>) => void;
  onSubmit: (values: Values, formikActions: FFActions<Values>) => void;
  validationSchema?: any | (() => any);
  validate?: ((values: Values) => void | object | Promise<FFErrors<Values>>);
}
export type FFErrors<Values> = FormikErrors<Values> & {};
export type FFSharedConfig = FormikSharedConfig & {};
export type FFState<Values> = FormikState<Values> & {};
export type FFActions<Values> = FormikActions<Values> & {};
export type FFHandlers = FormikHandlers & {
  handleClick(e: React.ChangeEvent<any>): void;
  handleClick<T = string | React.ChangeEvent<any>>(field: T): T;
};
export type FFComputedProps<Values> = FormikComputedProps<Values> & {};
export type FFRegistration = FormikRegistration & {};

export type FFProps<Values> = FFSharedConfig &
  FFState<Values> &
  FFActions<Values> &
  FFHandlers &
  FFComputedProps<Values> &
  FFRegistration;

export declare type FFContext<Values> = FFProps<Values> &
  Pick<FFConfig<Values>, 'validate' | 'validationSchema'>;
export interface SharedRenderProps<T> {
  component?: string | React.ComponentType<T | void>;
  render?: ((props: T) => React.ReactNode);
  children?: ((props: T) => React.ReactNode);
}

export declare type GenericFieldHTMLAttributes =
  | React.InputHTMLAttributes<HTMLInputElement>
  | React.SelectHTMLAttributes<HTMLSelectElement>
  | React.TextareaHTMLAttributes<HTMLTextAreaElement>;
