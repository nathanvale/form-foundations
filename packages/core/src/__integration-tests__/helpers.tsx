import * as React from 'react';
import { render, RenderResult } from 'react-testing-library';
import { withFF } from '../';
import { FFProps } from '../types';

// tslint:disable-next-line:no-empty
export const noop = () => {};

export const sleep = (ms: number) =>
  new Promise(resolve => setTimeout(resolve, ms));

export function renderWithFF<Values = {}>(
  Form: React.SFC<FFProps<Values>>,
  initialValues: any,
  options?: any,
  props?: any,
): RenderResult & { getProps: () => any } {
  let injected: any;

  const Component = props => (injected = props) && <Form {...props} />;
  Component.displayName = 'FFForm';

  const FFForm = withFF<{}, Values>({
    mapPropsToValues: () => initialValues,
    handleSubmit: noop,
    ...options,
  })(Component);

  expect(FFForm.displayName).toBe(
    'WithFormik(withState(withHandlers(mapProps(withProvider))))',
  );

  return {
    getProps() {
      return injected;
    },
    ...render(<FFForm {...props} />),
  };
}
