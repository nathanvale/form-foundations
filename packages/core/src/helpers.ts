import { TextFieldProps } from '@form-foundations/atoms';
import { FormikErrors, FormikTouched, getIn } from 'formik';
import { FieldPropsInjected, TextFieldNormalizeProps } from './types';

/**
 *
 */
export function callAll(...fns) {
  return (...args) => fns.forEach(fn => fn && fn(...args));
}

/**
 * This mapper abstracts away the tedious wrapper code needed to
 * convert formik field props to Material UI shapes. It is exported so
 * you can build on top of it (with functional composition) for more
 * specialized behavior.
 */
export function materialUIMapper({
  field,
  form,
  disabled = false,
  ...props
}: FieldPropsInjected<TextFieldProps>): any /*TODO: remove any */ {
  const { name } = field;
  const { touched, errors, isSubmitting } = form;
  const showError = isFieldShowingError({ errors, touched, name });
  return {
    ...props,
    ...field,
    error: showError,
    helperText: getFieldHelperText({
      errors: errors,
      name: name,
      helperText: props.helperText,
      touched,
    }),
    disabled: isSubmitting || disabled,
  };
}

/**
 * (value, previousValue)
 */
export function normalizeMapper({
  field,
  form,
  normalize,
  ...props
}: FieldPropsInjected<TextFieldNormalizeProps>): any /*TODO: remove any */ {
  const { name } = field;
  const { values } = form;
  const previousValue = values[name];
  field.onChange = event => {
    const { value } = event.target;
    form.setFieldValue(name, value ? normalize(value, previousValue) : '');
  };
  return {
    ...props,
    field,
    form,
  };
}

/**
 *
 */
export function fieldError(props: {
  errors: FormikErrors<{}>;
  name: string;
}): string {
  const { errors, name } = props;
  const fieldError = getIn(errors, name);
  return fieldError;
}

/**
 *
 */
export function getFieldHelperText({
  errors,
  name,
  touched,
  helperText,
}: {
  errors: FormikErrors<{}>;
  name: string;
  touched: FormikTouched<{}>;
  helperText: string | undefined;
}): string | undefined {
  return isFieldShowingError({ errors, name, touched })
    ? fieldError({ errors, name })
    : helperText;
}

/**
 *
 */
export function isFieldShowingError<Values = {}>({
  errors,
  touched,
  name,
}: {
  errors: FormikErrors<Values>;
  touched: FormikTouched<Values>;
  name: string;
}): boolean {
  return getIn(touched, name) && !!fieldError({ errors, name });
}
