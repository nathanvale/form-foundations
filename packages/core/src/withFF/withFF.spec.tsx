import * as React from 'react';
import { cleanup, fireEvent, wait } from 'react-testing-library';
import * as Yup from 'yup';
import { FFProps } from '../types';
import { renderWithFF } from '../__integration-tests__/helpers';

interface Values {
  firstName: string;
  lastName: string;
}

afterEach(cleanup);

const Form: React.SFC<FFProps<Values>> = ({
  values,
  handleSubmit,
  handleChange,
  handleBlur,
  handleClick,
  handleReset,
  touched,
  setStatus,
  status,
  errors,
  isSubmitting,
}) => {
  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        aria-label="first-name"
        onChange={handleChange}
        onBlur={handleBlur}
        onClick={handleClick}
        value={values.firstName}
        name="firstName"
      />
      {touched.firstName && errors.firstName && (
        <div id="feedback">{errors.firstName}</div>
      )}
      <input
        type="text"
        aria-label="last-name"
        onChange={handleChange}
        onBlur={handleBlur}
        onClick={handleClick}
        value={values.lastName}
        name="lastName"
      />
      {touched.lastName && errors.lastName && (
        <div id="feedback">{errors.lastName}</div>
      )}
      {isSubmitting && <div id="submitting">Submitting</div>}
      <button type="button" onClick={handleReset}>
        Reset
      </button>
      <button
        id="statusButton"
        onClick={() => setStatus({ myStatusMessage: 'True' })}
      >
        Call setStatus
      </button>
      {status && !!status.myStatusMessage && (
        <div id="statusMessage">{status.myStatusMessage}</div>
      )}
      <button type="submit">Submit</button>
    </form>
  );
};

const initialValues: Values = { firstName: 'Nathan', lastName: 'Vale' };

describe('withFF()', () => {
  it('should initialize Formik state and pass down props', () => {
    const { getProps } = renderWithFF<Values>(Form, initialValues);
    const props = getProps();
    expect(props).toMatchSnapshot();
  });

  it('should call render child element', () => {
    const { container } = renderWithFF<Values>(Form, initialValues);
    expect(container.firstChild).toBeDefined();
  });

  it('should call validate with values and props', async () => {
    const validate = jest.fn();
    const myProps = { my: 'prop' };
    const { getProps } = renderWithFF<Values>(
      Form,
      initialValues,
      { validate },
      myProps,
    );

    getProps().submitForm();
    await wait(() =>
      expect(validate).toHaveBeenCalledWith(
        { firstName: 'Nathan', lastName: 'Vale' },
        myProps,
      ),
    );
  });

  it('should call validationSchema', async () => {
    const validate = jest.fn(() => Promise.resolve());
    const { getProps } = renderWithFF<Values>(Form, initialValues, {
      validationSchema: { validate },
    });

    getProps().submitForm();
    await wait(() => expect(validate).toHaveBeenCalled());
  });

  it('should call validationSchema function with props', async () => {
    const validationSchema = jest.fn(() => Yup.object());
    const myProps = { my: 'prop' };
    const { getProps } = renderWithFF<Values>(
      Form,
      initialValues,
      {
        validationSchema,
      },
      myProps,
    );

    getProps().submitForm();
    await wait(() => expect(validationSchema).toHaveBeenCalledWith(myProps));
  });

  it('should call handleSubmit with values, actions and custom props', async () => {
    const handleSubmit = jest.fn();
    const myProps = { my: 'prop' };
    const { getProps } = renderWithFF<Values>(
      Form,
      initialValues,
      {
        handleSubmit,
      },
      myProps,
    );

    getProps().submitForm();

    await wait(() =>
      expect(handleSubmit).toHaveBeenCalledWith(
        { firstName: 'Nathan', lastName: 'Vale' },
        {
          props: myProps,
          resetForm: expect.any(Function),
          setError: expect.any(Function),
          setErrors: expect.any(Function),
          setFieldError: expect.any(Function),
          setFieldTouched: expect.any(Function),
          setFieldValue: expect.any(Function),
          setFormikState: expect.any(Function),
          setStatus: expect.any(Function),
          setSubmitting: expect.any(Function),
          setTouched: expect.any(Function),
          setValues: expect.any(Function),
          submitForm: expect.any(Function),
          validateField: expect.any(Function),
          validateForm: expect.any(Function),
        },
      ),
    );
  });

  it('should passe down custom props', () => {
    const { getProps } = renderWithFF<Values>(
      Form,
      initialValues,
      {},
      { my: 'prop' },
    );
    expect(getProps().my).toEqual('prop');
  });

  it('should set the currently active field', () => {
    const { getByLabelText, getProps } = renderWithFF<Values>(
      Form,
      initialValues,
      {},
    );
    let input = getByLabelText('first-name');
    fireEvent.click(input);
    let props = getProps();
    expect(props.active).toEqual({
      firstName: true,
    });
    input = getByLabelText('last-name');
    fireEvent.click(input);
    props = getProps();
    expect(props.active).toEqual({
      lastName: true,
    });
  });

  it('should reset the form', () => {
    const { getByLabelText, getByText, getProps } = renderWithFF<Values>(
      Form,
      initialValues,
      {},
    );
    let input = getByLabelText('first-name');
    fireEvent.click(input);
    fireEvent.blur(input);
    input = getByLabelText('last-name');
    fireEvent.click(input);
    fireEvent.blur(input);
    let button = getByText('Reset');
    fireEvent.click(button);
    let props = getProps();
    expect(props).toMatchSnapshot();
    expect(props.active).toEqual({});
  });
});
