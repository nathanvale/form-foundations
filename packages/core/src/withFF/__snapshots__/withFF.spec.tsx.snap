// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFF() should initialize Formik state and pass down props 1`] = `
Object {
  "active": Object {},
  "dirty": false,
  "errors": Object {},
  "formStarted": false,
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleClick": [Function],
  "handleReset": [Function],
  "handleSubmit": [Function],
  "initialValues": Object {
    "firstName": "Nathan",
    "lastName": "Vale",
  },
  "isSubmitting": false,
  "isValid": false,
  "isValidating": false,
  "registerField": [Function],
  "resetForm": [Function],
  "setError": [Function],
  "setErrors": [Function],
  "setFieldError": [Function],
  "setFieldTouched": [Function],
  "setFieldValue": [Function],
  "setFormikState": [Function],
  "setState": [Function],
  "setStatus": [Function],
  "setSubmitting": [Function],
  "setTouched": [Function],
  "setValues": [Function],
  "submitCount": 0,
  "submitForm": [Function],
  "touched": Object {},
  "unregisterField": [Function],
  "validateField": [Function],
  "validateForm": [Function],
  "validateOnBlur": true,
  "validateOnChange": true,
  "values": Object {
    "firstName": "Nathan",
    "lastName": "Vale",
  },
}
`;

exports[`withFF() should reset the form 1`] = `
Object {
  "active": Object {},
  "dirty": false,
  "error": undefined,
  "errors": Object {},
  "formStarted": false,
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleClick": [Function],
  "handleReset": [Function],
  "handleSubmit": [Function],
  "initialValues": Object {
    "firstName": "Nathan",
    "lastName": "Vale",
  },
  "isSubmitting": false,
  "isValid": false,
  "isValidating": false,
  "registerField": [Function],
  "resetForm": [Function],
  "setError": [Function],
  "setErrors": [Function],
  "setFieldError": [Function],
  "setFieldTouched": [Function],
  "setFieldValue": [Function],
  "setFormikState": [Function],
  "setState": [Function],
  "setStatus": [Function],
  "setSubmitting": [Function],
  "setTouched": [Function],
  "setValues": [Function],
  "status": undefined,
  "submitCount": 0,
  "submitForm": [Function],
  "touched": Object {},
  "unregisterField": [Function],
  "validateField": [Function],
  "validateForm": [Function],
  "validateOnBlur": true,
  "validateOnChange": true,
  "values": Object {
    "firstName": "Nathan",
    "lastName": "Vale",
  },
}
`;
